torch/csrc/autograd/python_function.cpp

```diff
@@ -257,7 +274,14 @@ static int THPFunction_clear(THPFunction* self) {
 
   self->output_info.clear();
   self->input_info.clear();
-  self->saved_variables.clear();
+  {
```

Comment:
cc @bdhirsh actually it was this doing bad things with memory and everything else was going bad.
Breaking during SavedVariable destruction before adding this shows both
```cpp
(gdb) bt
#0  0x00007fffe98b5200 in torch::autograd::SavedVariable::~SavedVariable() ()
   from /home/albandes/local/pytorch/3.11_debug_source/torch/lib/libtorch_python.so
#1  0x00007fffe98ad40c in THPFunction_clear(THPFunction*) ()
   from /home/albandes/local/pytorch/3.11_debug_source/torch/lib/libtorch_python.so
#2  0x00000000005257cd in subtype_clear (self=<CompiledFunctionBackward at remote 0x7fff0a745050>)
    at Objects/typeobject.c:1297
#3  0x00000000006435db in delete_garbage (tstate=tstate@entry=0x962618 <_PyRuntime+166328>, 
    gcstate=gcstate@entry=0x948520 <_PyRuntime+59584>, collectable=collectable@entry=0x7fffffffd0a0, 
    old=old@entry=0x948568 <_PyRuntime+59656>) at Modules/gcmodule.c:1013

```

and then

```cpp
(gdb) bt
#0  0x00007fffe98b5200 in torch::autograd::SavedVariable::~SavedVariable() ()
   from /home/albandes/local/pytorch/3.11_debug_source/torch/lib/libtorch_python.so
#1  0x00007fffe98b291c in torch::autograd::PyNode::release_variables() ()
   from /home/albandes/local/pytorch/3.11_debug_source/torch/lib/libtorch_python.so
#2  0x00007fffe600b212 in torch::autograd::deleteNode(torch::autograd::Node*) ()
   from /home/albandes/local/pytorch/3.11_debug_source/torch/lib/libtorch_cpu.so
#3  0x00007fffe98b432e in std::_Sp_counted_deleter<torch::autograd::PyNode*, void (*)(torch::autograd::Node*), std::allocator<void>, (__gnu_cxx::_Lock_policy)2>::_M_dispose() ()
   from /home/albandes/local/pytorch/3.11_debug_source/torch/lib/libtorch_python.so
#4  0x00007fffe5fec09a in torch::autograd::AutogradMeta::~AutogradMeta() ()
   from /home/albandes/local/pytorch/3.11_debug_source/torch/lib/libtorch_cpu.so
#5  0x00007fffea085e04 in c10::TensorImpl::~TensorImpl() ()
   from /home/albandes/local/pytorch/3.11_debug_source/torch/lib/libc10.so
#6  0x00007fffea085f19 in c10::TensorImpl::~TensorImpl() ()
   from /home/albandes/local/pytorch/3.11_debug_source/torch/lib/libc10.so
#7  0x00007fffe603e84c in torch::autograd::DifferentiableViewMeta::~DifferentiableViewMeta() ()
   from /home/albandes/local/pytorch/3.11_debug_source/torch/lib/libtorch_cpu.so
#8  0x00007fffea085e04 in c10::TensorImpl::~TensorImpl() ()
   from /home/albandes/local/pytorch/3.11_debug_source/torch/lib/libc10.so
#9  0x00007fffea085f19 in c10::TensorImpl::~TensorImpl() ()
   from /home/albandes/local/pytorch/3.11_debug_source/torch/lib/libc10.so
#10 0x00007fffe98ad40c in THPFunction_clear(THPFunction*) ()
   from /home/albandes/local/pytorch/3.11_debug_source/torch/lib/libtorch_python.so
#11 0x00000000005257cd in subtype_clear (self=<CompiledFunctionBackward at remote 0x7fff0a745050>)
    at Objects/typeobject.c:1297
#12 0x00000000006435db in delete_garbage (tstate=tstate@entry=0x962618 <_PyRuntime+166328>, 
    gcstate=gcstate@entry=0x948520 <_PyRuntime+59584>, collectable=collectable@entry=0x7fffffffd0a0, 
    old=old@entry=0x948568 <_PyRuntime+59656>) at Modules/gcmodule.c:1013

```

Not sure what's the expected behavior but that definitely fixes segfaults locally.